### DISCUSSION TITLE
Heltec SX1262

### DISCUSSION THREAD

**Discussion by nglessner** [timmbogner/Farm-Data-Relay-System#108] [Q&A] [answered]
I picked up a couple of [these](https://heltec.org/project/htcc-ab02/) boards from Heltec recently hoping to use them with your FDRS framework. I've already implemented a few ESPNow nodes and love the flexibility. From what I gather, the LoRa library that is used in fdrs_sensor.h (LoRa.h) is not compatible with the newer SX1262 boards, is that correct? My path forward would be to implement a "driver" for this Heltec board - something like USE_LORA_1262 right?

I'm pretty new to the microcontroller and github world, but thought I might make an attempt to contribute.

**Comment by timmbogner** [comment]
Awesome, thanks!
I've done some research and it looks like a worthy endeavor.
Yes, the way would be to add a condition after it checks for USE_LORA. You can leave out the lora part, and have it be "USE_SX1262". The else of that condition should then have the code for default usage. Hope that makes sense.
I checked out the BeeGee Tokyo library and it looks good, but the major difference will be that it uses callbacks to check for incoming data. I'll be happy to help if you need any.

**Reply by PerRieland** [reply]
I have some heltec lora 32 v3 boards, that I would like to use. These boards use SX1262. @nglessner did you manage to implement support for SX1262 ?

**Reply by timmbogner** [reply]
I have switched to the RadioLib library, which should make this addition quite simple now. It looks like the only API difference between the two is the name of the interrupt pin (DIO1 vs DIO0). See [SX126x](https://github.com/jgromes/RadioLib/blob/master/examples/SX126x/SX126x_PingPong/SX126x_PingPong.ino) vs [SX127x](https://github.com/jgromes/RadioLib/blob/master/examples/SX127x/SX127x_PingPong/SX127x_PingPong.ino).

I'm in the midst of a big code overhaul, and hopefully the new version will be a lot easier to work with, [check it out here](https://github.com/timmbogner/Farm-Data-Relay-System/tree/release_2023_01).

At some point I will make the small changes and post them here for you guys to test, if you don't mind!

**Reply by PerRieland** [reply]
Great to hear ;)
I would love to test it out on the SX1262 chip.
Keep up the good work

**Reply by timmbogner** [reply]
Cool! I just added it to the new release, which you can find in the pull requests section. It is enabled by default (for now), you'll just need to set your pins correctly: DIO refers to DIO1 on sx1262. The first thing I need to know is whether RadioLib initializes the chip successfully. Next, try using the radio by sending/receiving data to another LoRa unit that you know works, or else set up a second SX1262 gateway as a neighbor and confirm they are communicating. Thanks!

**Reply by timmbogner** [reply]
Okay actually I just realized the CubeCell series isn't ESP-based. I will have to do some more things to support ASR605x. I'll look into it later tonight... it may or may not be a big project. Tagging @nglessner in case you weren't notified.

**Reply by timmbogner** [reply]
I just found the schematic for the Heltec LoRa32v3 and set the pin configurations, so I think it should work for you now without changes @PerRieland.

**Reply by PerRieland** [reply]
Thanks a lot for working on the SX1262 chip, I appreciate it very much. 
I have been trying to get it to work on the Heltec board (https://heltec.org/project/wifi-lora-32-v3/) . I had to make some small adjustment to get it compile. Here is my feedback.
The Gateway can compile and upload. It connect to WiFi, it connect to MQTT. The serial monitor show MQTT messages from MQTT server to Gateway. It' running without crashing.  :) I'm a happy man

Next is the Node. Here I have problems.
It can compile, upload fine. When it's run, It keeps crashing/rebooting. The boards works just fine with factory code from Heltec.
If I comment out "addFDRS(1000, fdrs_recv_cb)" then it's not crashing, but also not working :(

I upload the sketch for mqttgateway and node. Maybe it gives you a clue what's wrong.  
My next step is to order some Lora modules from another manufactory, to see if it solves the problem.

I think it has to do with interrupt ??

18:27:22.416 -> Rebooting...
18:27:22.416 -> ESP-ROM:esp32s3-20210327
18:27:22.416 -> Build:Mar 27 2021
18:27:22.416 -> rst:0xc (RTC_SW_CPU_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
18:27:22.463 -> Saved PC:0x4200c963
18:27:22.463 -> SPIWP:0xee
18:27:22.463 -> mode:DIO, clock div:1
18:27:22.463 -> load:0x3fce3808,len:0x44c
18:27:22.463 -> load:0x403c9700,len:0xbec
18:27:22.463 -> load:0x403cc700,len:0x2920
18:27:22.463 -> entry 0x403c98d8
18:27:24.572 -> Q
18:27:24.572 -> Guru Meditation Error: Core  1 panic'ed (IllegalInstruction). Exception was unhandled.
18:27:24.572 -> Memory dump at 0x420018d8: 024fe5f9 00000090 21004136
18:27:24.572 -> Core  1 register dump:
18:27:24.572 -> PC      : 0x420018dc  PS      : 0x00060830  A0      : 0x3fc94c94  A1      : 0x3fcec2a0  
18:27:24.572 -> A2      : 0x000007d0  A3      : 0x00000000  A4      : 0x00000078  A5      : 0x0000002c  
18:27:24.604 -> A6      : 0x0000002b  A7      : 0x0000e100  A8      : 0x820018dc  A9      : 0x3fcec280  
18:27:24.604 -> A10     : 0x00000003  A11     : 0x00000001  A12     : 0x00001d0f  A13     : 0x00001021  
18:27:24.604 -> A14     : 0x00000001  A15     : 0x00000012  SAR     : 0x00000010  EXCCAUSE: 0x00000000  
18:27:24.604 -> EXCVADDR: 0x00000000  LBEG    : 0x4037cf30  LEND    : 0x4037cf3e  LCOUNT  : 0x00000001  
18:27:24.648 -> 
18:27:24.648 -> 
18:27:24.648 -> Backtrace: 0x420018d9:0x3fcec2a0 0x3fc94c91:0x3fcec2c0 |<-CORRUPTED

Below are the adjustment I have made to make it compile and run

fdrs_gateway_lora.h: Line 78
//Per RADIOLIB_MODULE radio = new Module(LORA_SS, LORA_DIO, LORA_RST, -1);
SX1262 radio = new Module(LORA_SS, LORA_DIO, LORA_RST, LORA_BUSY);

fdrs_gateway_lora.h: Line 244
#ifdef USE_LORA
void begin_lora()
{
//Per A
   SPI.begin(LORA_SCK,LORA_MISO,LORA_MOSI,LORA_SS);
  //NOT WORKING. int state = radio.begin(FDRS_LORA_FREQUENCY, FDRS_LORA_BANDWIDTH, FDRS_LORA_SF, FDRS_LORA_CR, FDRS_LORA_SYNCWORD, FDRS_LORA_TXPWR, 8, 0);
  //OK WORKING int state = radio.begin();
  //OK WORKING
  int state = radio.begin(FDRS_LORA_FREQUENCY, FDRS_LORA_BANDWIDTH, FDRS_LORA_SF, FDRS_LORA_CR, FDRS_LORA_SYNCWORD, FDRS_LORA_TXPWR, 8, 1.6, false);
//Per A 

************

fdrs_node.h: Line 114
#ifdef USE_LORA
//Per RADIOLIB_MODULE radio = new Module(LORA_SS, LORA_DIO, LORA_RST, -1);
RADIOLIB_MODULE radio = new Module(LORA_SS, LORA_DIO, LORA_RST, LORA_BUSY);


fdrs_node.h: Line 228
#ifdef USE_LORA
  //Per A
   SPI.begin(LORA_SCK,LORA_MISO,LORA_MOSI,LORA_SS);
  //NOT WORKING. int state = radio.begin(FDRS_LORA_FREQUENCY, FDRS_LORA_BANDWIDTH, FDRS_LORA_SF, FDRS_LORA_CR, FDRS_LORA_SYNCWORD, FDRS_LORA_TXPWR, 8, 0);
  //OK WORKING int state = radio.begin();
  //OK WORKING
  int state = radio.begin(FDRS_LORA_FREQUENCY, FDRS_LORA_BANDWIDTH, FDRS_LORA_SF, FDRS_LORA_CR, FDRS_LORA_SYNCWORD, FDRS_LORA_TXPWR, 8, 1.6, false);
//Per A
  if (state == RADIOLIB_ERR_NONE) {
    DBG("RadioLib initialization successful!");
  } else {
    DBG("RadioLib initialization failed, code " + String(state));
    while (true);
  }
  //Per B
  DBG("LoRa Initialized. Frequency: " + String(FDRS_LORA_FREQUENCY) + "  Bandwidth: " + String(FDRS_LORA_BANDWIDTH) + "  SF: " + String(FDRS_LORA_SF) + "  CR: " + String(FDRS_LORA_CR) + "  SyncWord: " + String(FDRS_LORA_SYNCWORD) + "  Tx Power: " + String(FDRS_LORA_TXPWR) + "dBm");
  #ifdef USE_SX126X
      radio.setDio1Action(setFlag);
  #else
      radio.setDio0Action(setFlag);
  #endif
//Per B

************

fdrs_gateway_config.h
//  FARM DATA RELAY SYSTEM
//
//  GATEWAY 2.000 Configuration

#include <fdrs_globals.h>

#define FDRS_DEBUG     //Enable USB-Serial debugging
#define DEBUG_CONFIG // Displays full config info on startup

#define UNIT_MAC     0x00  // The address of this gateway

// Actions -- Define what happens when a packet arrives at each interface:
// Current function options are: sendESPNow(MAC), sendSerial(), sendMQTT(),
//  sendLoRaNbr(interface) , sendESPNowNbr(interface).

#define ESPNOWG_ACT    
#define SERIAL_ACT     
#define MQTT_ACT       sendLoRaNbr(1); 
#define LORAG_ACT      sendMQTT();    

// Protocols -- Define which protocols the gateway will use.
// Warning: ESP-NOW and WiFi should not be used simultaneously.

//#define USE_ESPNOW  
#define USE_LORA
#define USE_WIFI    //Used only for MQTT gateway
#define USE_SX126X

// Neighbor addresses
#define ESPNOW_NEIGHBOR_1 0x0E  // ESPNOW1 Address 
#define ESPNOW_NEIGHBOR_2 0x0F  // ESPNOW2 Address
#define LORA_NEIGHBOR_1    0x0E  // LoRa1 Address
#define LORA_NEIGHBOR_2    0x0F  // LoRa2 Address

// Neighbor Actions
#define ESPNOW1_ACT    
#define ESPNOW2_ACT                    
#define LORA1_ACT 
#define LORA2_ACT 

// LoRa Configuration
#define RADIOLIB_MODULE SX1262 //Tested on SX1276
#define LORA_SS 8
#define LORA_RST 12
#define LORA_DIO 14
#define LORA_TXPWR 17   // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))
#define LORA_ACK        // Request LoRa acknowledgment.
#define LORA_BUSY 13
#define LORA_SCK 9
#define LORA_MISO 11
#define LORA_MOSI 10

************

Single_Channel_Relay_Lora_TEST.ino: 

#define READING_ID 71
#define GTWY_MAC 0x00
#define USE_LORA
#define COIL_PIN 48

#define USE_SX126X
// LoRa Configuration
#define RADIOLIB_MODULE SX1262 //Tested on SX1276
#define LORA_SS 8
#define LORA_RST 12
#define LORA_DIO 14
#define LORA_TXPWR 17   // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))
#define LORA_ACK        // Request LoRa acknowledgment.
#define LORA_BUSY 13
#define LORA_SCK 9
#define LORA_MISO 11
#define LORA_MOSI 10


#include <fdrs_node.h>

bool status = 0;

void fdrs_recv_cb(DataReading theData) {
  status = (bool)theData.d;
}
void setup() {
  Serial.begin(115200);
  beginFDRS();
  delay(2000);
  if (addFDRS(1000, fdrs_recv_cb))
    subscribeFDRS(READING_ID);
  pinMode(COIL_PIN, OUTPUT);
}
void loop() {
  if (status) digitalWrite(COIL_PIN, HIGH);
  else digitalWrite(COIL_PIN, LOW);
  DBG("COIL_PIN " + String(COIL_PIN));
  Serial.println("Q");
  // delay(3000);
}


[FDRS-2023-01.zip](https://github.com/timmbogner/Farm-Data-Relay-System/files/10677904/FDRS-2023-01.zip)

**Reply by timmbogner** [reply]
Thanks for testing it! Sorry I was unclear, but the node firmware has not been updated, only the gateway. This means no LoRa controllers or SX1262 nodes, although it will be added shortly. My intent here was to just confirm my implementation on the gateway was successful before I copy it over to nodes. Nodes will end up working a little differently, and I think addFDRS() will be phased out of usage. Basically: LoRa nodes won't need to ask the gateway to send them broadcasts, because the gateway will be sending the data to everybody regardless. 
What should work right now:
- LoRa Sensor (with SX127x) to -> LoRa Gateway (with SX1262)
- LoRa Gateway (with SX1262) <- to -> LoRa Gateway (with SX1262 or SX127x)

*With all of that said...* we first need to find out why it won't compile for you out of the box. For me, in Arduino v2.0.3, with esp32 board definitions v2.0,5, compiling "1_UART_Gateway.ino" for "Heltec WiFi LoRa 32 (V2)" is successful. What error messages do you get when you try to compile?
EDIT to add: Also update to the latest RadioLib if you haven't.

**Reply by timmbogner** [reply]
@PerRieland @nglessner I just added in the support for SX126x nodes. Also: I also was finally able to order a Heltec LoRa32 v3, so I will be able to give this a try as soon as it arrives (7-12 days).

**Comment by nglessner** [comment]
I didn't, I've been working on some ESPNow stuff lately.

If I remember correctly, it should be very simple. As you said, it's just a
matter of implementing the callback.

**Comment by nglessner** [comment]
They do have some that are esp32-s3 with sx1262. Their website isn't the
easiest to see what all of their options are.

**Reply by timmbogner** [reply]
I see now... I was confusing the two posts. I'd say to just avoid the CubeCell boards at this point. Hopefully the LoRa32v3 is successful.

**Comment by nglessner** [comment]
Fantastic! I've been busy with other projects lately, but will be
revisiting LoRa later this spring.

**Comment by timmbogner** [comment, marked as answer]
I'm very pleased to announce that SX1262 support is now tested and confirmed!


#


You must enable ```USE_SX126X``` in addition to setting ```#define RADIOLIB_MODULE SX1262``` and your pins.

The correct configuration for the Heltec LoRa32 v3 is as follows:
``` cpp
// LoRa Configuration
#define RADIOLIB_MODULE SX1262
#define LORA_SS 8
#define LORA_RST 12
#define LORA_DIO 14
#define USE_SX126X
```
**Thanks to everyone who suggested this, it's been a long time coming!**

**Comment by timmbogner** [comment]
There's a bit of a hiccup in this that I didn't catch apparently. The controller and ping functionality isn't working under SX1262. Strangely, it does receive ACKs.

I will look at this in the coming few days.

**Reply by machasm1** [reply]
Hi Tim,
I am trying to use this board https://www.waveshare.com/wiki/Pico-LoRa-SX1262-868M#Overview with fdrs and I am quite confused as to which pins to wire up. When using pin numbers as stated in your post above which pins are you referring to?
I am also looking at driving these modules from a raspberry pico eventually but happy to use an esp8266 for now.
The GPIO pins on the esp8266 or actual pin numbers running from top to bottom? I am probably missing something really obvious!
Any help/guidance would be great.

**Reply by timmbogner** [reply]
Here is a diagram. The pins you use are your choice, but they must be mapped to the following:

![image](https://user-images.githubusercontent.com/64260873/222921203-adb947a4-ee5d-4f97-832f-43ff170d2340.png)

The SPI pins that aren't marked on this should be attached to[ the chip's default SPI pins](https://randomnerdtutorials.com/esp8266-pinout-reference-gpios/). The ~~BUSY pin~~ and battery voltage pin can be ignored.

Let me know how it goes! I'll also be curious if a Pico with LoRa would work on FDRS... It compiled okay last I checked.

**Reply by timmbogner** [reply]
@aviateur17 I was just looking at the pin usage on your Hallard LoRa Wemos shield to see what 8266 pins might be recommended. I'm still unclear... can you advise a good set of pins to use on 8266?

**Comment by machasm1** [comment]
Connected to Pico using the header and configured fdrs_node_config.h as such

// LoRa Configuration
//#define RADIOLIB_MODULE SX1276 //Tested on SX1276
#define RADIOLIB_MODULE SX1262
#define USE_SX126X
#define LORA_SS 3
#define LORA_RST 15
#define LORA_DIO 20

#define LORA_TXPWR 17    // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))
#define LORA_ACK        // Request LoRa acknowledgment.

//#define CUSTOM_SPI
#define LORA_SPI_SCK 10
#define LORA_SPI_MISO 12
#define LORA_SPI_MOSI 11

Although the code did compile I am getting an initialization error as below when monitoring the serial port after uploading the .uf2 file

Message (Enter to send message to 'Raspberry Pi Pico W' on 'COM15')
Both NL & CR
9600 baud
RadioLib initialization failed, code -2

I am not sure if it is a compatability problem with RadioLib.h? Since I have got the device to work when using semtechs libraries but not with fdrs

**Reply by timmbogner** [reply]
For starters it's that Pi handles SPI a little differently and some changes were needed. I think I might have found the solution, but we'll never know because after the code uploaded to my Pi, it stopped responding completely. It won't open a COM port and just makes random USB connections and disconnections. This is after a bunch of issues with the Pico just not cooperating well with Arduino in general. I think I'm just going to say that Pi Pico support for Arduino is currently less than optimal, and leave it there. Let's try to use that radio with an ESP instead.

**Comment by aviateur17** [comment]
I would try:


// LoRa Configuration
//#define RADIOLIB_MODULE SX1276 //Tested on SX1276
#define RADIOLIB_MODULE SX1262
#define USE_SX126X
#define LORA_SS 3
#define LORA_RST 15
#define LORA_DIO 2

#define LORA_TXPWR 17 // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))
#define LORA_ACK // Request LoRa acknowledgment.

#define CUSTOM_SPI
#define LORA_SPI_SCK 10
#define LORA_SPI_MISO 12
#define LORA_SPI_MOSI 11
 
I think the only difference is that I would uncomment the #define CUSTOM_SPI line and the LORA_DIO change from pin 20 to pin 2.
Uncommenting the CUSTOM_SPI line just makes sure that the pins you define below that line are used for the SPI communication to the chip.   The reason for the LORA_DIO change is that the LoRa chips normally have DIO0, DIO1, DIO2, DIO3, etc for some predefined general purpose I/O.   DIO1 is defined as Pin 20 but we want to use DIO0 and not DIO1.  I believe the LoRa_BUSY pin is actually DIO0 so I would define LoRa_BUSY which is Pin 2 to DIO0.

**Reply by timmbogner** [reply]
I'm actually working on it too and at exactly this point. The problem is that I have instances for setting the custom pins for ESP32 and ESP8266, but I'll need to add whatever method Pi uses, as well as find out what ifdef the PI uses.

**Reply by timmbogner** [reply]
SX127x DIO0 is actually equivalent to SX126x DIO1
 SX127x DIO1 is equivalent to SX126x BUSY
It's confusing.  DIO20 was right. 
I think that this would be the Pi setup code:
```
  SPI.setRX(LORA_SPI_MISO);
  SPI.setTX(LORA_SPI_MOSI);
  SPI.setSCK(LORA_SPI_SCK);
  SPI.begin();
```
However, uploading the test code somehow bricked my Pi Pico, so I'm kinda done messing with it.

**Reply by machasm1** [reply]
As far as unbricking the pico try this
https://github.com/dwelch67/raspberrypi-pico/blob/main/flash_nuke.uf2
Upload the flash_nuke.uf2 file to the pico in the normal way.

Back to the problem at hand,
when I uncomment the #define CUSTOM_SPI line I get the following compilation error.

In file included from c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h:70,
                 from C:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\examples\LoRa_Sensor\LoRa_Sensor.ino:10:
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h: In function 'crcResult handleLoRa()':
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h:122:13: error: 'radio' was not declared in this scope
  122 |             radio.startReceive(); // return to listen mode
      |             ^~~~~
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h:131:17: error: 'radio' was not declared in this scope
  131 |                 radio.startReceive();
      |                 ^~~~~
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h: In function 'void begin_lora()':
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h:149:15: error: 'radio' was not declared in this scope
  149 |   int state = radio.begin(FDRS_LORA_FREQUENCY, FDRS_LORA_BANDWIDTH, FDRS_LORA_SF, FDRS_LORA_CR, FDRS_LORA_SYNCWORD, FDRS_LORA_TXPWR);
      |               ^~~~~
In file included from c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h:70,
                 from C:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\examples\LoRa_Sensor\LoRa_Sensor.ino:10:
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h: In function 'crcResult transmitLoRa(uint16_t*, DataReading*, uint8_t)':
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h:221:21: error: 'radio' was not declared in this scope
  221 |         int state = radio.transmit(pkt, sizeof(pkt));
      |                     ^~~~~
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h: In function 'crcResult transmitLoRa(uint16_t*, SystemPacket*, uint8_t)':
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h:305:17: error: 'radio' was not declared in this scope
  305 |     int state = radio.startTransmit(pkt, sizeof(pkt));
      |                 ^~~~~
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h: In function 'crcResult getLoRa()':
c:\Users\joe\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node_lora.h:326:22: error: 'radio' was not declared in this scope
  326 |     int packetSize = radio.getPacketLength();
      |                      ^~~~~

exit status 1

Compilation error: exit status 1

**Reply by timmbogner** [reply]
Somehow I got the Pico back to being able to talk to Arduino again, and when tried the sketch again the same thing happens: COM port disappears and an error comes up about 20 seconds after plugging it in that says "USB Device Not Recognized" (on Windows 10). If you want to test drive this , the [changes and branch are here.](https://github.com/timmbogner/Farm-Data-Relay-System/commit/6f40c85f9f9e9840f8e2636c938b7e8357383ec4) Fair warning, though...

I have it wired to an SX1278 module using the same pins listed in your module's docs.

**Reply by timmbogner** [reply]
[continued here](https://github.com/timmbogner/Farm-Data-Relay-System/commit/6f40c85f9f9e9840f8e2636c938b7e8357383ec4)

**Comment by machasm1** [comment]
Tried the updated branch and ran into similar problems as yourself although the script would now compile ok.
I was, however, able to rejuvenate my pico by first nuking it using flash_nuke.uf2 then loading any other code onto it. I just used the Lora Sensor node without the modified code and it reverted to behaving as it did before so no long term damage.
Just a hunch but perhaps the modified code somehow might be using the same resources that the usb interface on the pico needs to communicate with the pc??

**Comment by nglessner** [comment]
Is the sketch using deep sleep? I had this experience with the Seeed
Studios Xiao C3 with deep sleep. Basically the usb to serial connection is
turned off when the device is in deep sleep, which caused the USB
disconnect.
